### PR TITLE
Bugfix/singleton resumption quickfix

### PR DESCRIPTION
**Pull request by m-samiec**
#### Description
Enable singletons to resume from snapshots.

#### Release note
Singletons authority and state resumes correct when reconnecting servers to snapshot.

#### Tests
Tested locally with single and multiple servers and DW validated.

#### Primary reviewers
@Vatyx

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 1]
```diff
@@ -1275,6 +1279,25 @@ TWeakObjectPtr<USpatialActorChannel> USpatialReceiver::PopPendingActorRequest(Wo
...
+	UGameplayStatics::GetAllActorsOfClass(NetDriver->World, SingletonClass, FoundActors);
+
+	// There should be only one singleton actor per class
+	if (FoundActors.Num() == 1)
+	{
+		return FoundActors[0];
+	}
+	else
+	{
+		UE_LOG(LogSpatialReceiver, Warning, TEXT("Found incorrect number (%d) of singleton actors (%s)"),
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: I remember we discussed maybe way back when this should be a crash or something? Or nah? Or an error?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp [hunk 2]
```diff
@@ -363,6 +363,17 @@ USpatialActorChannel* UGlobalStateManager::AddSingleton(AActor* SingletonActor)
 	return Channel;
 }
 
+void UGlobalStateManager::RegisterSingletonChannel(AActor* SingletonActor, USpatialActorChannel* SingletonChannel)
+{
+	TPair<AActor*, USpatialActorChannel*>& ActorChannelPair = NetDriver->SingletonActorChannels.FindOrAdd(SingletonActor->GetClass());
```
[the remark this answers is not in the document]
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp]: You can setup the pair, before you have an instanced actor available - LinkAllExistingSingletonActors

[the remark this answers is not in the document]
Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp]: Amazing

**Comment by Vatyx** [comment]
dude sick pr bruh

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 1 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Could update to error.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 3]
```diff
@@ -1275,6 +1279,25 @@ TWeakObjectPtr<USpatialActorChannel> USpatialReceiver::PopPendingActorRequest(Wo
 	return Channel;
 }
 
+AActor* USpatialReceiver::FindSingletonActor(UClass* SingletonClass)
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: FindSingletonActor seems to be generic enough to not be in USpatialReceiver. What if we move it to NetDriver instead?

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: It's not a method you'd want to encourage to use, `GetAllActorsOfClass` is expensive, and only used here as a last resort.

Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: I agree, my motivation is that it is more of a utility function and not a part of SpatialReceiver. I'm fine with it staying in SpatialReceiver though.

**Pull request merged**